net/redeemer.proto

```diff
@@ -0,0 +1,34 @@
+syntax = "proto3";
+
+import "google/protobuf/empty.proto";
+
+package net;
+
+service TicketRedeemer {
+    rpc QueueTicket(Ticket) returns (google.protobuf.Empty) {}
+    rpc MaxFloat(MaxFloatRequest) returns (MaxFloatUpdate) {}
+    rpc MonitorMaxFloat(google.protobuf.Empty) returns (stream MaxFloatUpdate) {}
```

Comment:
Same goes for the use of `google.protobuf.Empty` here - let's stub out an empty request parameter.

(Perhaps I don't quite understand yet how the pieces here come together, but wouldn't we want to parameterize MonitorMaxFloat by the broadcaster address?)

Reply:
> (Perhaps I don't quite understand yet how the pieces here come together, but wouldn't we want to parameterize MonitorMaxFloat by the broadcaster address?)

That would mean a listener for each sender rather than a single listener for all updates though.

Reply:
Is it an issue to have a listener for each sender? Having a single listener for all updates means each orchestrator node gets updates from all senders across the entire multi-O fleet, even though a particular orchestrator doesn't need the majority of that data. That's pretty chatty and will have a lower scalability ceiling than a listener-per-sender.

Reply:
I believe we can still use a single listener and the service as is. 

The difference between sending a message for a `sender` to all clients instead of to only clients that are working with that sender would simply require additional bookkeeping on the server using a map that can be cleaned up on the same interval as the `liveSenders` map. 

We would have to alter `MonitorMaxFloat` to take in a `message MonitorMaxFlotReq{ bytes sender = 1; }`. 

The functionality would be as it is now , if there is already a stream existing for that client we will just use that stream to push updates. Additionally we will now also add the sender to the map for which that client wants to receive updates.

Reply:
Looked at this some more

> even though a particular orchestrator doesn't need the majority of that data. That's pretty chatty and will have a lower scalability ceiling than a listener-per-sender.

There's nothing chatty about it in my opinion, it's one way traffic. It's essentially a gossip protocol which scales very well with hundreds of clients and hundreds of messages. We'll never even reach such scale because it pertains to _winning_ tickets and a _multi-O_ setup. An O controls its winning probability and amount of clients in the cluster. If you have high win probability for tickets you should alter your parameters, if you have many O's in your fleet, consider running a multi-T setup instead.  

I also think it's not entirely correct to say they don't need the majority of the data. They _might_ not need it , but they might as well need it due to load balancing: 

- Client side load balancing: All O's supplied with `orchAddr` will need to discover the max float for the broadcaster, not pushing updates to all clients means doing remote calls.

- Server side load balancing:  If the request gets directed to another node during the transcode workflow we now have to do a remote request on that other node to get the maxfloat for B even though the cluster was already aware. 

All in all I think only pushing particular updates to clients is a premature optimisation and leads to possible synchronisation issues that need to be aleviated with remote calls.

Reply:
In general, writing to the network is a bit like writing to the blockchain: avoid if possible.  But I *think* I can see the reasoning for doing that here, especially coupled with this note https://github.com/livepeer/go-livepeer/pull/1521#discussion_r432424537 .

A few more questions, just to clarify my own understanding:

> Client side load balancing: All O's supplied with orchAddr will need to discover the max float for the broadcaster, not pushing updates to all clients means doing remote calls.

* You mean `redeemerAddr` ?
* Wouldn't a remote call still be necessary in most cases, in order to obtain the max float during discovery? How is the load balanced here?

> Server side load balancing: If the request gets directed to another node during the transcode workflow we now have to do a remote request on that other node to get the maxfloat for B even though the cluster was already aware.

* That is true only if a winning ticket has been submitted by the first node? In most cases, a remote call would still be necessary for pricing info?

So altogether, it seems like we're only saving ourselves RPC calls for the (relatively infrequent) case where a winning ticket is received? Anyway, for now, keeping things relatively stateless seems okay. Just ask that we also incorporate https://github.com/livepeer/go-livepeer/pull/1513#discussion_r431337263 so we have room to make modifications later if necessary.

Please let me know if the following flow is correct (it's a bit hard to piece everything together from the various PRs, and I don't see a canonical description / doc anywhere):

1. On startup, O subscribes to `MonitorMaxFloat` on a RPC redeemer for float updates
1. During discovery / GetOrchestratorInfo , an O invokes MaxFloat RPC and caches the resulting value for the given B. Only that one O gets the MaxFloat value. (How long is the value kept in cache?)
1. Anytime a winning ticket is submitted to the redeemer, a MaxFloat update for a given B is broadcast to all O subscribers, whether or not they're working with the B. (Does this also create a local cached value for the max float even if the current orchestrator isn't working with the B?)

BTW it would be great to have this behavior documented somewhere, eg in `/doc`, including the DB schema updates.

Reply:
Will definitely document the behaviour as I notice it is a source of confusion. 

Fixed the initial comment from this thread in 258e02747160df22b465787adfd245ce9ed02a65